### ISSUE TITLE
Installation issue: scipy with macOS Sonoma/Apple Clang 15

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#40351]
### Steps to reproduce the issue

```console
> spack spec -I py-scipy
Input spec
--------------------------------
 -   py-scipy

Concretized
--------------------------------
 -   py-scipy@1.11.3%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]      ^openblas@0.3.24%apple-clang@15.0.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=darwin-sonoma-m2
[+]          ^perl@5.38.0%apple-clang@15.0.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=darwin-sonoma-m2
[+]              ^berkeley-db@18.1.40%apple-clang@15.0.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=darwin-sonoma-m2
[+]      ^pkgconf@1.9.5%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m2
[+]          ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]      ^py-cython@0.29.36%apple-clang@15.0.0 build_system=python_pip patches=c4369ad arch=darwin-sonoma-m2
[+]          ^py-setuptools@63.4.3%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]      ^py-meson-python@0.13.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]          ^meson@1.2.1%apple-clang@15.0.0 build_system=python_pip patches=0f0b1bd arch=darwin-sonoma-m2
[+]          ^py-pyproject-metadata@0.7.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]              ^py-packaging@23.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]                  ^py-flit-core@3.9.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]      ^py-numpy@1.25.2%apple-clang@15.0.0+blas+lapack build_system=python_pip patches=873745d arch=darwin-sonoma-m2
[+]      ^py-pip@23.1.2%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]      ^py-pybind11@2.10.4%apple-clang@15.0.0~ipo build_system=cmake build_type=Release generator=ninja arch=darwin-sonoma-m2
[+]          ^cmake@3.27.6%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=darwin-sonoma-m2
[+]              ^curl@8.1.2%apple-clang@15.0.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=secure_transport arch=darwin-sonoma-m2
[+]                  ^nghttp2@1.52.0%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m2
[+]          ^ninja@1.11.1%apple-clang@15.0.0+re2c build_system=generic arch=darwin-sonoma-m2
[+]              ^re2c@2.2%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]      ^py-pythran@0.12.2%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]          ^llvm-openmp@16.0.0%apple-clang@15.0.0~ipo~multicompat build_system=cmake build_type=Release generator=make arch=darwin-sonoma-m2
[+]              ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=autotools arch=darwin-sonoma-m2
[+]          ^py-beniget@0.4.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]          ^py-gast@0.5.3%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]          ^py-ply@3.11%apple-clang@15.0.0 build_system=python_pip arch=darwin-sonoma-m2
[+]      ^py-wheel@0.37.1%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]      ^python@3.11.4%apple-clang@15.0.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis+optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=darwin-sonoma-m2
[e]          ^apple-libuuid@1353.100.2%apple-clang@15.0.0 build_system=bundle arch=darwin-sonoma-m2
[+]          ^bzip2@1.0.8%apple-clang@15.0.0~debug~pic+shared build_system=generic arch=darwin-sonoma-m2
[+]              ^diffutils@3.9%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m2
[+]          ^expat@2.5.0%apple-clang@15.0.0~libbsd build_system=autotools arch=darwin-sonoma-m2
[+]          ^gdbm@1.23%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m2
[+]          ^gettext@0.21.1%apple-clang@15.0.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=darwin-sonoma-m2
[+]              ^libiconv@1.17%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-sonoma-m2
[+]              ^libxml2@2.10.3%apple-clang@15.0.0+pic~python+shared build_system=autotools arch=darwin-sonoma-m2
[+]              ^tar@1.34%apple-clang@15.0.0 build_system=autotools zip=pigz arch=darwin-sonoma-m2
[+]                  ^pigz@2.7%apple-clang@15.0.0 build_system=makefile arch=darwin-sonoma-m2
[+]                  ^zstd@1.5.5%apple-clang@15.0.0+programs build_system=makefile compression=none libs=shared,static arch=darwin-sonoma-m2
[+]          ^libffi@3.4.4%apple-clang@15.0.0 build_system=autotools arch=darwin-sonoma-m2
[+]          ^libxcrypt@4.4.35%apple-clang@15.0.0~obsolete_api build_system=autotools patches=4885da3 arch=darwin-sonoma-m2
[+]          ^ncurses@6.4%apple-clang@15.0.0~symlinks+termlib abi=none build_system=autotools arch=darwin-sonoma-m2
[+]          ^openssl@3.1.3%apple-clang@15.0.0~docs+shared build_system=generic certs=mozilla arch=darwin-sonoma-m2
[+]              ^ca-certificates-mozilla@2023-05-30%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m2
[+]          ^readline@8.2%apple-clang@15.0.0 build_system=autotools patches=bbf97f1 arch=darwin-sonoma-m2
[+]          ^sqlite@3.42.0%apple-clang@15.0.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=darwin-sonoma-m2
[+]          ^xz@5.4.1%apple-clang@15.0.0~pic build_system=autotools libs=shared,static arch=darwin-sonoma-m2
[+]          ^zlib-ng@2.1.3%apple-clang@15.0.0+compat+opt build_system=autotools patches=299b958,86c3f5b,ae9077a,b692621 arch=darwin-sonoma-m2
```

### Error message

<details><summary>Error message</summary>

<pre>
  [298/1619] Linking target scipy/special/_ellip_harm_2.cpython-311-darwin.so
  FAILED: scipy/special/_ellip_harm_2.cpython-311-darwin.so
  /Users/Adam/spack/lib/spack/env/clang/clang  -o scipy/special/_ellip_harm_2.cpython-311-darwin.so scipy/special/_ellip_harm_2.cpython-311-darwin.so.p/meson-generated__ellip_harm_2.c.o scipy/special/_ellip_harm_2.cpython-311-darwin.so.p/sf_error.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,dynamic_lookup -bundle -Wl,-undefined,dynamic_lookup -lm -Wl,-rpath,/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/openblas-0.3.24-lsn4jcy34p62ysj7qq2ghlsawu6sco6a/lib /Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/openblas-0.3.24-lsn4jcy34p62ysj7qq2ghlsawu6sco6a/lib/libopenblas.dylib
  ld: warning: duplicate -rpath '/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/openblas-0.3.24-lsn4jcy34p62ysj7qq2ghlsawu6sco6a/lib' ignored
  ld: duplicate LC_RPATH '/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/gcc-13.2.0-qhn4xdjhn2p7k2fnvllfz2ca6xhwl2yu/lib' in '/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/openblas-0.3.24-lsn4jcy34p62ysj7qq2ghlsawu6sco6a/lib/libopenblas-r0.3.24.dylib'
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
</pre></details>

### Information on your system

* **Spack:** 0.21.0.dev0 (e59665217e74093c18080a99575f20622836da6f)
* **Python:** 3.11.4
* **Platform:** darwin-sonoma-m2
* **Concretizer:** clingo

This is on the #40057 branch.

### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/12831121/spack-build-out.txt)
* [spack-build-env-mods.txt](https://github.com/spack/spack/files/12831122/spack-build-env-mods.txt)

@rgommers let me know if you want me to open an issue on the scipy GitHub too.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by rgommers** [comment]
Hmm, that `LC_RPATH` thing is supposed to be a warning rather than a hard error from everything I just read. A few links:
- [https://github.com/igraph/igraph/issues/2394 indicates it's](https://github.com/igraph/igraph/issues/2394#issuecomment-1734139366) shows this happening with `libopenblas` and Fortran runtime libraries that are relevant for SciPy
- https://indiestack.com/2023/10/xcode-15-duplicate-library-linker-warnings/ is a useful write-up of what's up with this new behavior

The failure in `ellip_harm_2` may not be related to that warning though, it may fail to link for a completely unrelated reason (it's unusual code, for example there's a try-finally in Cython code around a call to a C callback mechanism). The linker shipped in XCode 15 is completely new apparently, so we're probably hitting a corner case here.

> @rgommers let me know if you want me to open an issue on the scipy GitHub too.

That may be useful, yes. There are no changes to that pyx file since the last release, so I suspect this issue will show up elsewhere and not just in Spack. I may not be able to look into it for the next week or so, pretty swamped with other stuff.

**Comment by adamjstewart** [comment]
Closed by #40057 which forces use of the classic linker
